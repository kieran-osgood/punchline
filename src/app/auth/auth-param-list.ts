import { StackNavigationProp } from '@react-navigation/stack';
import { RouteProp } from '@react-navigation/native';

export type RouteParamList = {
  Login: undefined;
  Register: undefined;
};

export type AuthNavigationProps<T extends keyof RouteParamList> = {
  route: RouteProp<RouteParamList, T>;
  navigation: StackNavigationProp<RouteParamList, T>;
};
