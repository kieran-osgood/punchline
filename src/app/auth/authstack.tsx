import React from 'react';
import { createStackNavigator } from '@react-navigation/stack';
import { Button, Text } from 'react-native';

import { AuthNavigationProps, RouteParamList } from './authparamlist';
import CenterView from '../../components/centerview';

const Stack = createStackNavigator<RouteParamList>();

function Login({ navigation }: AuthNavigationProps<'Login'>) {
  return (
    <CenterView>
      <Text>I am a login screen</Text>
      <Button
        title="log me in"
        onPress={() => {
          console.log('test');
        }}
      />
      <Button
        title="go to register"
        onPress={() => {
          navigation.navigate('Register');
        }}
      />
    </CenterView>
  );
}

function Register({ navigation, route }: AuthNavigationProps<'Register'>) {
  return (
    <CenterView>
      <Text>route name: {route.name}</Text>
      <Button
        title="go to login"
        onPress={() => {
          navigation.navigate('Login');
          // navigation.goBack()
        }}
      />
    </CenterView>
  );
}

export const AuthStack = () => {
  return (
    <Stack.Navigator
      screenOptions={{
        header: () => null,
      }}
      initialRouteName="Login">
      <Stack.Screen
        options={{
          headerTitle: 'Sign In',
        }}
        name="Login"
        component={Login}
      />
      <Stack.Screen
        options={{
          headerTitle: 'Sign Up',
        }}
        name="Register"
        component={Register}
      />
    </Stack.Navigator>
  );
};
