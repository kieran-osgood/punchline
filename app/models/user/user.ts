import { FirebaseAuthTypes } from "@react-native-firebase/auth"
import { Instance, SnapshotOut, types } from "mobx-state-tree"
import { NonFunctionProperties } from "types"
// FIXME: rename this to firebase user to match what the intent is
type UserDataType = NonFunctionProperties<FirebaseAuthTypes.User>
export const UserData: UserDataType = {
  displayName: "",
  email: "",
  emailVerified: false,
  isAnonymous: false,
  metadata: {
    creationTime: "",
    lastSignInTime: "",
  },
  phoneNumber: "",
  photoURL: "",
  providerData: [],
  providerId: "",
  uid: "",
}

// Used for R.pick to extract properties from the FirebaseAuthTypes.UserCredential
export const UserDataKeys = Object.keys(UserData)

/**
 * Currently Authenticated Users Firebase Information .
 */
export const UserModel = types
  .model("User")
  .props({
    displayName: types.maybeNull(types.string),
    email: types.maybeNull(types.string),
    emailVerified: types.boolean,
    isAnonymous: types.boolean,
    metadata: types.model({
      creationTime: types.maybe(types.string),
      lastSignInTime: types.maybe(types.string),
    }),
    phoneNumber: types.maybeNull(types.string),
    photoURL: types.maybeNull(types.string),
    providerData: types.frozen(),
    providerId: types.string,
    uid: types.string,
  })
  .views((self) => ({})) // eslint-disable-line @typescript-eslint/no-unused-vars
  .actions((self) => ({})) // eslint-disable-line @typescript-eslint/no-unused-vars

/**
 * Un-comment the following to omit model attributes from your snapshots (and from async storage).
 * Useful for sensitive data like passwords, or transitive state like whether a modal is open.

 * Note that you'll need to import `omit` from ramda, which is already included in the project!
 *  .postProcessSnapshot(omit(["password", "socialSecurityNumber", "creditCardNumber"]))
 */

type UserType = Instance<typeof UserModel>
export interface User extends UserType {}
type UserSnapshotType = SnapshotOut<typeof UserModel>
export interface UserSnapshot extends UserSnapshotType {}
export const createUserDefaultModel = () => types.optional(UserModel, UserData)
