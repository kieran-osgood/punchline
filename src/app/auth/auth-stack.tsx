import React from 'react';
import { createStackNavigator } from '@react-navigation/stack';

import { RouteParamList } from './auth-param-list';
import Register from './screens/register';
import LoginChoices from './screens/login-choices';
import EmailPassword from './screens/email-password';

const Stack = createStackNavigator<RouteParamList>();

export const AuthStack = () => {
  return (
    <Stack.Navigator
      // screenOptions={{
      //   header: () => null,
      // }}
      initialRouteName="LoginChoices">
      <Stack.Screen
        options={{
          headerTitle: 'Email / Password',
        }}
        name="EmailPassword"
        component={EmailPassword}
      />
      <Stack.Screen
        options={{
          headerTitle: 'Login',
        }}
        name="LoginChoices"
        component={LoginChoices}
      />
      <Stack.Screen
        options={{
          headerTitle: 'Sign Up',
        }}
        name="Register"
        component={Register}
      />
    </Stack.Navigator>
  );
};
