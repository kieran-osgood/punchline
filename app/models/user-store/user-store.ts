import { FirebaseAuthTypes } from "@react-native-firebase/auth"
import { JokeLength } from "app/graphql"
import { RootStore } from "app/models"
import { cast, getRoot, Instance, SnapshotOut, types } from "mobx-state-tree"
import R from "ramda"
import { withEnvironment } from "../extensions/with-environment"
import { UserDataKeys, UserModel } from "../user/user"

export const UserStoreModel = types
  .model("UserStore", {
    user: types.maybeNull(UserModel),
    jokeLengthPreferences: types.map(types.boolean),
  })
  .extend(withEnvironment)
  .views((self) => ({
    get blockedCategoryIds(): string[] {
      const root = [...(getRoot(self) as RootStore).api.categories.values()]
        .filter((x) => x.isActive)
        .map((x) => x.id)
      return root
    },
    get jokeLengthMax(): number {
      let largestJoke = 0
      self.jokeLengthPreferences.forEach((x, b) => {
        if (x) {
          switch (b) {
            case JokeLength.LARGE:
              largestJoke = largestJoke <= 3 ? 3 : largestJoke
            case JokeLength.MEDIUM:
              largestJoke = largestJoke <= 2 ? 2 : largestJoke
            case JokeLength.SMALL:
              largestJoke = largestJoke <= 1 ? 1 : largestJoke
          }
        }
      })

      return largestJoke
    },
  }))
  .actions((self) => ({
    updateUser: (user: FirebaseAuthTypes.User | null) => {
      if (user === null) {
        self.user = null
        return
      }
      self.user = cast(R.pick(UserDataKeys, user))
    },
    toggleJokeLength: (value: JokeLength, isChecked?: boolean) => {
      self.jokeLengthPreferences.set(value, isChecked ?? false)
    },
  }))

type UserStoreType = Instance<typeof UserStoreModel>
export interface UserStore extends UserStoreType {}
type UserStoreSnapshotType = SnapshotOut<typeof UserStoreModel>
export interface UserStoreSnapshot extends UserStoreSnapshotType {}
export const createUserStoreDefaultModel = () => types.optional(UserStoreModel, {})
