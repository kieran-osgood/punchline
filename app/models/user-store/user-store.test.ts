import { FirebaseAuthTypes } from "@react-native-firebase/auth"
import { UserData, UserModel } from "app/models/user/user"
import { UserStoreModel } from "./user-store"

test("can be created without any user", () => {
  const instance = UserStoreModel.create({})
  expect(instance).toMatchSnapshot()
})

test("can be created with user", () => {
  const user = UserModel.create(UserData)
  const store = UserStoreModel.create({ user })

  expect(store).toMatchSnapshot()
})

const FirebaseUserCredential: FirebaseAuthTypes.User = {
  displayName: "display me",
  email: "email@gmail.com",
  emailVerified: false,
  isAnonymous: true,
  metadata: {
    creationTime: "1619118612167",
    lastSignInTime: "1619118612167",
  },
  phoneNumber: null,
  photoURL: null,
  providerData: [],
  providerId: "firebase",
  uid: "89PMk9NNPnMyv7fuoQMhKTjZRiA2",
  updateProfile: jest.fn(),
  verifyBeforeUpdateEmail: jest.fn(),
  delete: jest.fn(),
  getIdToken: jest.fn(),
  getIdTokenResult: jest.fn(),
  linkWithCredential: jest.fn(),
  reauthenticateWithCredential: jest.fn(),
  reload: jest.fn(),
  sendEmailVerification: jest.fn(),
  unlink: jest.fn(),
  toJSON: jest.fn(),
  updateEmail: jest.fn(),
  updatePassword: jest.fn(),
  updatePhoneNumber: jest.fn(),
}

test("can updateUser with FirebaseAuthTypes.User object", () => {
  const user = UserModel.create(UserData)
  const store = UserStoreModel.create({ user })
  store.updateUser(FirebaseUserCredential)
  expect(store).toMatchSnapshot()
})
