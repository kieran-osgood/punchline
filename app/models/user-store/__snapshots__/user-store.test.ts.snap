// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be created with user 1`] = `
Object {
  "user": Object {
    "displayName": "",
    "email": "",
    "emailVerified": false,
    "isAnonymous": false,
    "metadata": Object {
      "creationTime": "",
      "lastSignInTime": "",
    },
    "phoneNumber": "",
    "photoURL": "",
    "providerId": "",
    "uid": "",
  },
}
`;

exports[`can be created without any user 1`] = `
Object {
  "user": null,
}
`;

exports[`can updateUser with FirebaseAuthTypes.User object 1`] = `
Object {
  "user": Object {
    "displayName": "display me",
    "email": "email@gmail.com",
    "emailVerified": false,
    "isAnonymous": true,
    "metadata": Object {
      "creationTime": "1619118612167",
      "lastSignInTime": "1619118612167",
    },
    "phoneNumber": null,
    "photoURL": null,
    "providerId": "firebase",
    "uid": "89PMk9NNPnMyv7fuoQMhKTjZRiA2",
  },
}
`;
